tox_docker/tox4/plugin.py

```diff
@@ -0,0 +1,122 @@
+from typing import Dict, List
+
+from tox.config.cli.parser import ToxParser
+from tox.config.main import Config
+from tox.config.sets import ConfigSet
+from tox.config.types import EnvList
+from tox.execute.api import Outcome
+from tox.plugin import impl
+from tox.tox_env.api import ToxEnv
+
+from tox_docker.config import ContainerConfig, RunningContainers
+from tox_docker.plugin import (
+    docker_health_check,
+    docker_pull,
+    docker_run,
+    get_env_vars,
+    HealthCheckFailed,
+    stop_containers,
+)
+from tox_docker.tox4.config import (
+    discover_container_configs,
+    EnvRunningContainers,
+    parse_container_config,
+)
+from tox_docker.tox4.log import log
+
+CONTAINER_CONFIGS: Dict[str, ContainerConfig] = {}
+ENV_CONTAINERS: EnvRunningContainers = {}
+
+
+@impl
+def tox_add_core_config(core_conf: ConfigSet, config: Config) -> None:
+    container_config_names = discover_container_configs(config)
+
+    # validate command line options
+    for container_name in config.options.docker_dont_stop:
+        if container_name not in container_config_names:
+            raise ValueError(
+                f"Container {container_name!r} not found (from --docker-dont-stop)"
+            )
+
+    for container_name in container_config_names:
+        CONTAINER_CONFIGS[container_name] = parse_container_config(
+            config, container_name, container_config_names
+        )
+
+
+@impl
+def tox_before_run_commands(tox_env: ToxEnv) -> None:
+    tox_env.conf.add_config(
```

Comment:
Don't add config here, you should do it during the env config register phase (different plugin). This means that ``tox4 config`` command will not work as expected.

Reply:
This is still incorrect. You're only allowed to register configuration within `tox_add_core_config` and `tox_add_env_config`. I'll need to freeze the config set, so people can't do it later. https://github.com/tox-dev/tox/issues/2243

I've also put in https://github.com/tox-dev/tox/issues/2244 to add support for config sets to render correctly for `tox config`.

Reply:
I see. That wasn't really at all clear. I think a lot of work could be done to make the docs clear for plugin authors... I've had to figure a lot of this by trial and error.